### DISCUSSION TITLE
Possible to restrict rpc for policy security definer functions?

### DISCUSSION THREAD

**Discussion by AJamesPhillips** [supabase/supabase#3269] [Questions] [answered]
In the [docs for auth](https://supabase.io/docs/guides/auth#policies-with-security-definer-functions) to solve RLS for many-to-many relationships it gives an example of using a function created with `security definer` privileges.

```sql
create or replace function get_teams_for_user(user_id uuid)
returns setof bigint as $$
    select team_id
    from members
    where user_id = $1
$$ stable language sql security definer;
```

However this is callable from the client via rpc which reveals what teams (team ids) someone is a member of.  Is there a way to keep this information private?

**Comment by AJamesPhillips** [comment, marked as answer]
Actually it looks like it's [possible to rewrite this function](https://stackoverflow.com/a/69267166/539490) as:

```sql
create or replace function get_teams_for_user()
returns setof bigint as $$
    select team_id
    from members
    where user_id = auth.uid()
$$ stable language sql security definer;
```

Which then prevents other users from finding the teams other users are members of.  Is there a downside to this approach?

**Reply by AJamesPhillips** [reply]
It would still be useful to be able to define functions that can not be called directly via rpc from a client and only indirectly called by another function.  Perhaps this is already possible with Postgres grants / roles etc?

**Reply by bnjmnt4n** [reply]
Currently, Postgrest only exposes tables/functions/views in the `public` schema, so creating a new schema will hide it from the public Postgrest API, and I believe you can then modify the roles such that the function can be used in RLS policies/triggers etc.

**Reply by steve-chavez** [reply]
@AJamesPhillips As @bnjmnt4n pointed out, you can `create schema private` and then `create or replace function private.get_teams_for_user()..`, this way the function won't be callable through rpc.
